Frameworks/Foundation/NSOperationQueue.mm

```diff
@@ -281,64 +304,70 @@ - (void)observeValueForKeyPath:(NSString*)keyPath ofObject:(id)object change:(NS
  @Status Interoperable
 */
 + (NSOperationQueue*)currentQueue {
-    NSThread* currentThread = [NSThread currentThread];
-    if ([currentThread isMainThread]) {
+    if ([[NSThread currentThread] isMainThread]) {
         return [[self class] mainQueue];
     }
 
-    return [[currentThread threadDictionary] objectForKey:_NSOperationQueueCurrentQueueKey];
+    return tls_currentQueue.get();
 }
 
 /**
  @Status Interoperable
 */
 + (NSOperationQueue*)mainQueue {
-    static StrongId<NSOperationQueue> mainQueue = [[NSOperationQueue_MainQueue new] autorelease];
+    static StrongId<NSOperationQueue> mainQueue = [[_NSOperationQueue_MainQueue new] autorelease];
     return mainQueue;
 }
 
 /**
  @Status Interoperable
 */
 - (void)addOperation:(NSOperation*)op {
-    if (op.isExecuting || op.isFinished) {
-        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"operation is already or finished executing" userInfo:nil];
-    }
+    @synchronized(op) {
+        if (op.executing || op.finished) {
+            @throw [NSException exceptionWithName:NSInvalidArgumentException
+                                           reason:@"operation is already or finished executing"
+                                         userInfo:nil];
+        }
 
-    if (!op._acquirePermissionToAddToQueue) {
-        @throw [NSException exceptionWithName:NSInvalidArgumentException reason:@"operation is already in an operation queue" userInfo:nil];
-    }
+        if (!op._markInQueue) {
+            @throw [NSException exceptionWithName:NSInvalidArgumentException
+                                           reason:@"operation is already in an operation queue"
+                                         userInfo:nil];
+        }
 
-    [self willChangeValueForKey:@"operations"];
-    [self willChangeValueForKey:@"operationCount"];
+        [self willChangeValueForKey:@"operations"];
+        [self willChangeValueForKey:@"operationCount"];
 
-    @synchronized(_operations.get()) {
-        [_operations addObject:op];
-    }
+        @synchronized(_operations.get()) {
+            [_operations addObject:op];
+        }
 
-    [self didChangeValueForKey:@"operationCount"];
-    [self didChangeValueForKey:@"operations"];
+        [self didChangeValueForKey:@"operationCount"];
+        [self didChangeValueForKey:@"operations"];
 
-    [op addObserver:self forKeyPath:@"isFinished" options:0 context:nil];
-    [op addObserver:self forKeyPath:@"isReady" options:0 context:nil];
+        [op addObserver:self forKeyPath:@"isFinished" options:NSKeyValueObservingOptionNew context:_NSOperationQueue_IsFinishedContext];
```

Comment:
oops oops
i should version control a little more aggressively so i avoid these sorts of mistakes